internal/config/pipeline.go

```diff
@@ -0,0 +1,283 @@
+					DisablePush bool `hcl:"disable_push,optional"`
+				}
+
+				if diag := gohcl.DecodeBody(step.Use.Body, finalizeContext(c.ctx), &buildBody); diag.HasErrors() {
+					return nil, diag
+				}
+
+				s.Kind = &pb.Pipeline_Step_Build_{
+					Build: &pb.Pipeline_Step_Build{
+						DisablePush: buildBody.DisablePush,
+					},
+				}
+			case "deploy":
+				var deployBody struct {
+					Release bool `hcl:"release,optional"`
+				}
+
+				if diag := gohcl.DecodeBody(step.Use.Body, finalizeContext(c.ctx), &deployBody); diag.HasErrors() {
+					return nil, diag
+				}
+
+				s.Kind = &pb.Pipeline_Step_Deploy_{
+					Deploy: &pb.Pipeline_Step_Deploy{
+						Release: deployBody.Release,
+					},
+				}
+			case "release":
+				var releaseBody struct {
+					DeploymentRef       uint64 `hcl:"deployment_ref,optional"` // 0 or "unset" means latest
+					Prune               bool   `hcl:"prune,optional"`
+					PruneRetain         int32  `hcl:"prune_retain,optional"`
+					PruneRetainOverride bool   `hcl:"prune_retain_override,optional"`
+				}
+
+				if diag := gohcl.DecodeBody(step.Use.Body, finalizeContext(c.ctx), &releaseBody); diag.HasErrors() {
+					return nil, diag
+				}
+
+				// For parsing pipeline configs, an unset `deployment_ref` translates
+				// to the "latest" deployment. Otherwise if people want to release
+				// a specific deployment by sequence number, they can set it explicity.
+				deployRef := &pb.Ref_Deployment{
+					Ref: &pb.Ref_Deployment_Latest{
+						Latest: true,
+					},
+				}
+				if releaseBody.DeploymentRef != 0 {
+					deployRef = &pb.Ref_Deployment{
+						Ref: &pb.Ref_Deployment_Sequence{
+							Sequence: releaseBody.DeploymentRef,
+						},
+					}
+				}
+
+				s.Kind = &pb.Pipeline_Step_Release_{
+					Release: &pb.Pipeline_Step_Release{
+						Deployment:          deployRef,
+						Prune:               releaseBody.Prune,
+						PruneRetain:         releaseBody.PruneRetain,
+						PruneRetainOverride: releaseBody.PruneRetainOverride,
+					},
+				}
+			case "up":
+				var upBody struct {
+					Prune               bool  `hcl:"prune,optional"`
+					PruneRetain         int32 `hcl:"prune_retain,optional"`
+					PruneRetainOverride bool  `hcl:"prune_retain_override,optional"`
+				}
+
+				if diag := gohcl.DecodeBody(step.Use.Body, finalizeContext(c.ctx), &upBody); diag.HasErrors() {
+					return nil, diag
+				}
+
+				s.Kind = &pb.Pipeline_Step_Up_{
+					Up: &pb.Pipeline_Step_Up{
+						Prune:               upBody.Prune,
+						PruneRetain:         upBody.PruneRetain,
+						PruneRetainOverride: upBody.PruneRetainOverride,
+					},
+				}
+			case "exec":
+				var execBody struct {
+					Command string   `hcl:"command,optional"`
+					Args    []string `hcl:"args,optional"`
+				}
+
+				// Evaluate the step body hcl to get options
+				if diag := gohcl.DecodeBody(step.Use.Body, finalizeContext(c.ctx), &execBody); diag.HasErrors() {
+					return nil, diag
+				}
+
+				s.Kind = &pb.Pipeline_Step_Exec_{
```

Comment:
⏳ For a "step plugin" to be supported in pipelines, does it need to be added to the server proto? I'm thinking about how that could be extracted to the plugin SDK so any plugin could be used here, if so, & where in this package/elsewhere in "core" would need to be changed to enable that.

Reply:
That is correct. We intentionally didn't do the work to support external step plugins yet because we were thinking we're pretty far off from users wanting custom step plugins. But this is a good catch/good eye, we should think about what it'd take to do this when we're ready for it!